hack/monitorgen/main.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2021 The Cluster Monitoring Operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	metricsprofiles "github.com/openshift/cluster-monitoring-operator/hack/monitorgen/metrics"
+	"github.com/openshift/cluster-monitoring-operator/pkg/manifests"
+	"sigs.k8s.io/yaml"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+)
+
+const (
+	scrapeProfileLabel = "monitoring.openshift.io/collection-profile"
+)
+
+var (
+	path            *string
+	pathMetricsFile *string
+	// Map that contains ServiceMonitor names as keys and a list of metrics as
+	// value
+	minimalProfileMetrics = map[string][]string{
+		"kubelet":            metricsprofiles.KubeletMinimal,
+		"etcd":               metricsprofiles.EtcdMinimal,
+		"node-exporter":      metricsprofiles.NodeExporterMinimal,
+		"kube-state-metrics": metricsprofiles.KubeStateMetricsMinimal,
+		"prometheus-adapter": metricsprofiles.PrometheusAdapterMinimal,
+	}
+)
+
+func init() {
+	path = flag.String("path", "./assets/control-plane/service-monitor-kubelet.yaml", "(Required) Path to ServiceMonitor that we want to generate collection profiles")
+	pathMetricsFile = flag.String("metrics-file", "", "(Optional) Path to file containing a list of metrics that we want to the ServiceMonitor to Keep.")
+}
+
+func main() {
+
+	flag.Parse()
+
+	f, err := os.ReadFile(*path)
+	if err != nil {
+		log.Fatalf("failed to read asset %v: %e", *path, err)
+	}
+
+	m, err := manifests.NewServiceMonitor(bytes.NewReader(f))
+	if err != nil {
+		log.Fatalf("failed to convert to service monitor: %e", err)
+	}
+
+	// The script is only inteded to be used with resources of kind ServiceMonitor
+	if m.Kind != "ServiceMonitor" {
+		return
+	}
+
+	// Don't process ServiceMonitors that were generated by this tool as they will be overwritten
+	if val, ok := m.ObjectMeta.Annotations[scrapeProfileLabel]; ok && val != "full" {
+		return
+	}
+
+	// Get sub set of metrics to keep for the ServiceMonitor name based on the 
+	// hard coded value in minimalProfileMetrics 
+	metricsToKeep := minimalProfileMetrics[m.ObjectMeta.Name]
+
+	// Overwrite with metrics passed as argument by the user
+	if *pathMetricsFile != "" {
+		metricsToKeep, err = readMetricsFromFile(*pathMetricsFile)
+		if err != nil {
+			return
+		}
+	}
+
+	if len(metricsToKeep) == 0 {
+		return
+	}
```

Comment:
Shouldn't this part be in `generateCollectionProfileMinimal`? When we add another profile, we'd want to add a function `generateCollectionProfileNewProfile` to main no?

Reply:
Reworked this part of the code